cmd/gitserver/internal/git/cli/clibackend.go

```
@@ -0,0 +1,223 @@
+// corrupt on disk.
+const gitConfigMaybeCorrupt = "sourcegraph.maybeCorruptRepo"
+
+type ErrRepoCorrupted struct{ Stderr string }
+
+func (e ErrRepoCorrupted) Error() string {
+	return "repository is corrupted: " + e.Stderr
+}
+
+func checkMaybeCorruptRepo(ctx context.Context, logger log.Logger, git git.GitBackend, repo api.RepoName, stderr string) bool {
+	if !stdErrIndicatesCorruption(stderr) {
+		return false
+	}
+
+	logger = logger.With(log.String("repo", string(repo)), log.String("repo", string(repo)))
+	logger.Warn("marking repo for re-cloning due to stderr output indicating repo corruption", log.String("stderr", stderr))
+
+	// We set a flag in the config for the cleanup janitor job to fix. The janitor
+	// runs every minute.
+	err := git.Config().Set(ctx, gitConfigMaybeCorrupt, strconv.FormatInt(time.Now().Unix(), 10))
+	if err != nil {
+		logger.Error("failed to set maybeCorruptRepo config", log.Error(err))
+	}
+
+	return true
+}
+
+var (
+	// objectOrPackFileCorruptionRegex matches stderr lines from git which indicate
+	// that a repository's packfiles or commit objects might be corrupted.
+	//
+	// See https://github.com/sourcegraph/sourcegraph/issues/6676 for more
+	// context.
+	objectOrPackFileCorruptionRegex = lazyregexp.NewPOSIX(`^error: (Could not read|packfile) `)
+
+	// objectOrPackFileCorruptionRegex matches stderr lines from git which indicate that
+	// git's supplemental commit-graph might be corrupted.
+	//
+	// See https://github.com/sourcegraph/sourcegraph/issues/37872 for more
+	// context.
+	commitGraphCorruptionRegex = lazyregexp.NewPOSIX(`^fatal: commit-graph requires overflow generation data but has none`)
+)
+
+// stdErrIndicatesCorruption returns true if the provided stderr output from a git command indicates
+// that there might be repository corruption.
+func stdErrIndicatesCorruption(stderr string) bool {
+	return objectOrPackFileCorruptionRegex.MatchString(stderr) || commitGraphCorruptionRegex.MatchString(stderr)
+}
```

Corruption checks are now happening here in the CLI package, because depending on the implementation, the detection logic might need adjustment, for example if we use go-git in the future.

Instead of just logging corruption and returning a exec failed error, we return a special error type now, so that the caller can decide what to do about it (either just forward the error, or trigger some additional logic, like the corruption logging to the database).
After this logic ran, all we need to fixup corruption in the next janitor run is already done, so all behavior bolted on top of what this package does feels fairly optional, thus ok to me to break the concerns up here at this line.